ndinterp_capi/src/lib.rs

```diff
@@ -21,11 +24,14 @@ pub unsafe extern "C" fn create_cubic_interpolator1d(
     values_c: *const f64,
     size: usize,
 ) -> Box<grid::cubic::Cubic1d> {
-    let input = ArrayView1::from_shape_ptr(size, input_c);
+    // Use slice instead of vec so that rust doesn't take ownership of the data and releases
+    // the burden is on the function calling them
+    let slice_input = unsafe { slice::from_raw_parts(input_c, size) };
```

Comment:
Are you sure of this? Eventually, you're using `.to_vec()` anyhow, couldn't you just copy the array into the vector without direct use of an `unsafe` block?

Reply:
I can copy, but then the caveat is the same. And when passing a pointer with data in C++ I don't expect the the function getting hold of my data so I cannot use vec (and... maybe that means the unsafe is mandatory either way? this last part I'm not sure)

Reply:
`Vec` owns its data, while a slice doesn't. So you're already copying with `.to_vec()`, as documented in the standard library

https://doc.rust-lang.org/std/primitive.slice.html#method.to_vec

The only way out would be to make `Grid` holding a reference, but then you'd need to play with explicit lifetimes, and I'd not recommend.

So, my personal advice is just to copy in the most explicit way possible (or otherwise going all-in with references, but you will need to propagate it in the original library, making references-holding `struct`, instead of owning the data - but it's incredibly more complex).

Reply:
is it not better to leave the burden of how to do this to the compiler? How would a copy be better than using the standard library to say "hey, this is a slice of data, look at it" "now, make it yours as a vec".

(it's not retorical, maybe there's some subtlety I'm missing)

All-in with references for the sake of an external interface is a bad idea ofc. Imho external interfaces should take the burden of making the quantities agreeable to the original program without adding complexity there.

Reply:
> is it not better to leave the burden of how to do this to the compiler? How would a copy be better than using the standard library to say "hey, this is a slice of data, look at it" "now, make it yours as a vec".
> 
> (it's not retorical, maybe there's some subtlety I'm missing)

Standard library already has safe abstractions over `unsafe` operations, so the goal was to avoid the explicit use of `unsafe`.
You could be perfectly safe by yourself, as those who wrote the standard library. But you have to take care of the possible subtleties, and what you did consciously once, it might be unconsciously edited by someone else later on.
That's why, if possible, I'd try to avoid the manual dereference of the pointer (if the standard library already provides facilities for this, since converting an array to a `Vec` seems a common operation).

Reply:
However, the answer to the `unsafe` is that there is no way, because Rust can not guarantee that the pointer you're giving him is actually an array.

So, the only possible change would be to skip the intermediate slice, and directly go for a vector, since you're going to make a vector eventually in any case:
https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts

Reply:
But I'd change at least the comment, since you can not avoid copying.

Reply:
Ok, let me read about that a bit and I'll update at least the comment with my best understanding of it.

But basically you're telling me if I need an unsafe that means I should make an effort to find whether there's a way to avoid it (which makes sense).

Re: directly going for the vector, that one I'd like to explicitly avoid (actually, I'm not sure it would work for things like the C++ vector, because C++ would lose ownership of the data without knowing about it so you get the double clean I mentioned at some point)

Reply:
> Re: directly going for the vector, that one I'd like to explicitly avoid (actually, I'm not sure it would work for things like the C++ vector, because C++ would lose ownership of the data without knowing about it so you get the double clean I mentioned at some point)

Ok, I see your point. Even the standard library is mentioning it:

> Violating these may cause problems like corrupting the allocator’s internal data structures. [...] To avoid these issues, it is often preferable to do casting/transmuting using [slice::from_raw_parts](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html) instead.
> The ownership of ptr is effectively transferred to the Vec<T> which may then deallocate, reallocate or change the contents of memory pointed to by the pointer at will. Ensure that nothing else uses the pointer after calling this function.
> https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts

So, it's just fine to keep doing as you did, but the goal is exactly copying instead of moving. (Moreover, they are also talking of *ownership*, but it's at a different level than the usual one, since it's not tracked by the compiler).